packages/tokenbank_v1/contracts/modules/security/GuardianModule.sol

```diff
@@ -0,0 +1,184 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "../../lib/MathUint.sol";
+
+import "../../thirdparty/ERC1271.sol";
+
+import "../../iface/Wallet.sol";
+
+import "./SecurityModule.sol";
+
+
+/// @title GuardianModule
+contract GuardianModule is SecurityModule
+{
+    uint public pendingPeriod;
+    uint public confirmPeriod;
+
+    mapping (address => mapping(address => uint)) public pendingAdditions;
+    mapping (address => mapping(address => uint)) public pendingRemovals;
+
+    event GuardianAdditionPending   (address indexed wallet, address indexed guardian, uint confirmAfter);
+    event GuardianAdded             (address indexed wallet, address indexed guardian);
+    event GuardianAdditionCancelled (address indexed wallet, address indexed guardian);
+
+    event GuardianRemovalPending    (address indexed wallet, address indexed guardian, uint confirmAfter);
+    event GuardianRemoved           (address indexed wallet, address indexed guardian);
+    event GuardianRemovalCancelled  (address indexed wallet, address indexed guardian);
+
+    constructor(
+        SecurityStorage _securityStorage,
+        uint _pendingPeriod,
+        uint _confirmPeriod
+        )
+        public
+        SecurityModule(_securityStorage)
+    {
+        pendingPeriod = _pendingPeriod;
+        confirmPeriod = _confirmPeriod;
+    }
+
+    function addGuardian(
+        address wallet,
+        address guardian
+        )
+        external
+        nonReentrant
```

Comment:
Because `executeMetaTx` is `nonReentrant`, I guess all functions called by a metaTx can't be `nonReentrant`? Not really sure how to best fix this. Maybe if `msg.sender == address(this)` the reentrancy check is skipped?

Reply:
I added  `nonReentrantExceptFrom` and `nonReentrantExceptFromThis`

Reply:
Hmmm seems like there are some subtleties with this solution. With `nonReentrantExceptFrom` the `_guardValue` is set to `0` when exiting the function which would reset is, which means that afterwards calling a method like that you could actually do reentrancy again. So a function stack like this would be possible:
```
nonReentrant (_guardValue -> 1)
  nonReentrantExceptFrom (_guardValue -> 0 after function end)
  nonReentrant (_guardValue -> 1)
```

With some extra logic this could perhaps be solved, but it seems very error prone because it can be very subtle how things can go wrong.

I think it would probably be easier to just keep the normal `nonReentrant` everywhere (easy to understand), while making  `executeMetaTx` **not** `nonReentrant`. Instead,  `executeMetaTx` needs to follow the Checks-Effects-Interactions pattern that makes it inherently safe against reentrancy. I think that is currently already the case (`saveExecutedMetaTx` is done before the function call).

Reply:
Yeah, the previous implementation does have the bug. Fixing the modifiers is trivial, please let me know if you like the fix or not.

Reply:
I think it works correctly now, but functions now need to be a little bit aware of these more complex interactions which does make things more difficult and error-prone. I think just making ` executeMetaTx` not `nonReentrant` would be easier, but the current method with modifiers seems to be work fine as well.

Reply:
executeMetaTx should be nonReentrant. A couple of legit call stacks:

- `executeMetaTx` (in `moduleA`) calls` foo` (in `ModuleA`) which calls `bar` (in a `Wallet`) which calls `xyz` (in `moduleA`).


- `executeMetaTx` (in `moduleA`) calls` foo` (in `ModuleA`) which calls `bar` (in a `Wallet`) which calls `xyz` (in `moduleB`).

Reply:
`executeMetaTx` should be safe for reentrancy, but using `nonReentrant` is just one way to do it. Using the Checks-Effects-Interactions pattern is another way to do. And the benefit with the pattern method is that we don't modify the state of the contract, so we don't influence any other methods.

> * `executeMetaTx` (in `moduleA`) calls` foo` (in `ModuleA`) which calls `bar` (in a `Wallet`) which calls `xyz` (in `moduleA`).

I don't really see how a state-modifying method can be called back on a module (`xyz` in this example) that isn't a `view` method (view methods don't need to be `nonReentrant` so are irrelevant). I guess it's possible by using `transact`, but the `msg.sender` for these is the wallet, so contracts that are called can't reach the original module (unless it's a view method of course, which doesn't matter). And even if it would be possible to do it, I think `moduleA.foo` and `moduleA.xyz` wouldn't be safe against reentrancy even when using a variant of `nonReentrant` unless they both follow the Checks-Effects-Interactions pattern because they can be used for reentrancy by external contracts via the wallet and that could be dangerous.

Reply:
@Brechtpd please feel free to submit a PR for this :)

Reply:
Will do. :)